### PR TITLE
Server-side render: Test for accuracy of removeBlockSupportAttributes

### PR DESCRIPTION
**Pull request by mcsf**
See https://github.com/WordPress/gutenberg/pull/44491#issuecomment-1457977535

## Why?
In #44491, the `ServerSideRender` component gained a new `skipBlockSupportAttributes` prop used to strip blocks' attributes of block-supports-related attributes and styles. This is implemented in a local function, `removeBlockSupportAttributes`, in which those known attributes and styles were hardcoded.

I raised the concern in https://github.com/WordPress/gutenberg/pull/44491#issuecomment-1457977535 that this could easily get out of sync, as over time block-supports features are added or edited, since there is no obvious connection between that subsystem and the `ServerSideRender` component.

## What?
This PR explores a solution to that problem by adding unit tests dedicated to `removeBlockSupportAttributes`.

## How?
These tests are a bit atypical, since they they use Node's `fs` API to grep through source files in `lib/block-supports` to collect the names of attributes and styles that potentially correspond to block-supports features.

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for your efforts on this @mcsf 👍 

I appreciate this is still a work in progress but was just wondering if we might be able to leverage the theme.json schema to sanity-check known styles.

It has the `definitions.stylesProperties.properties` object that lists all but the special cases of anchor and layout supports. 

One catch with the current `STYLE_KEY_PATTERN` could be a block support that makes heavy use of `_wp_array_get` instead of `$block_attributes['style']['styleName']['...']` etc.

All in all, though, it would be great to provide move certainty that ServerSideRender is kept up to date with new block supports. There's already so many pieces to introducing a block support, missing an update somewhere is easy to do.

**Comment by mcsf** [comment]
Thanks for the feedback, @aaronrobertshaw!

> I appreciate this is still a work in progress but was just wondering if we might be able to leverage the theme.json schema to sanity-check known styles.
> 
> It has the `definitions.stylesProperties.properties` object that lists all but the special cases of anchor and layout supports.

This makes a lot more sense than the dirty thing I did. :)

**Comment by mcsf** [comment]
However, there are discrepancies:

found in `removeBlockSupportAttributes` | found in schema | found in both
-|-|-
-|-|border
-|-|color
-|css|-
elements|-|-
-|filter|-
layout|-|-
-|outline|-
position|-|-
-|-|shadow
-|-|spacing
-|-|typography

(schema sourced using `jq -r '.definitions.stylesProperties.properties | keys[]' < schemas/json/theme.json | sort`)
(comparison done with `comm`)

**Comment by aaronrobertshaw** [comment]
Thanks for the comparison breakdown, handy stuff. 👍 

I did a little further digging and here's what I found:

- **Elements**
    - Are a bit different as it generates styles for inner elements
    - I'm not sure if it has to be stripped from the ServerSideRender attributes or not.
    - I don't think it would hurt, maybe as an arbitrary addition to the disallow list.
- **Layout**
    - This has always been a bit of an outlier. It currently is only available as a setting.
    - There might be some work underway soon that would allow an individual block type to have its layout set under styles
    - Maybe it can be an arbitrary addition to the list to check
- **Position**
    - This one is a setting only within theme.json and won't make sense as a global style
    - So it won't be available under `definitions.stylesProperties.properties` but maybe we could plug some gaps via `settingsPropertiesComplete`
- **CSS**
    - Custom CSS is another special case I think. It could be anything and is more a last resort for the end user when our design tooling doesn't cover a need. I'd be inclined to not strip this one but don't hold that position strongly at all.
- **Filter**
    - I believe duotone is being reworked and might appear under `filter` in the future.
    - There are also other SVG filters being work on. 
    - Would it be better to lean towards removing the filter styles now or not?
- **Outline**
    - It appears this hasn't really been implemented as a block support
    - A partial implementation was added that allowed outlines to be set via theme.json
    - With no UI at either the global styles or individual block level, their won't be anything to strip from an individual block's style object
- **Dimensions**
    - Turns out the min-height block support uses `_wp_array_get` and so was missed by `STYLE_KEY_PATTERN`
    - It should have been added to the theme.json schema 
    - I spoke to @andrewserong about this and the position support, we'll get a PR merged that fixes up the theme.json
    
In summary, maybe the theme.json schema could still be useful if we get the schema fixed up and perhaps combine the styles with either an arbitrary list for the special cases like layout & position, or the settings properties.

This all highlights how fraught the process of adding block supports is currently. Having a single source of truth would be a welcome addition I think.

**Comment by mcsf** [comment]
Hey, @aaronrobertshaw, thanks a lot for the sharing those details. I just got back from a meetup and I'm catching up.

> In summary, maybe the theme.json schema could still be useful if we get the schema fixed up and perhaps combine the styles with either an arbitrary list for the special cases like layout & position, or the settings properties.
> 
> This all highlights how fraught the process of adding block supports is currently. Having a single source of truth would be a welcome addition I think.

Looking at this with fresh eyes, I am now of the mind that this exploration is too costly for the payoff: we are talking about a niche prop (`skipBlockSupportAttributes`) of a niche component (`ServerSideRender`) for blocks that, for legacy reasons, chose to not have a proper `edit` method. In practice, this affects blocks Archive and Tag Cloud; should any of these start rendering poorly, anyone working on them can now trace the prop `skipBlockSupportAttributes` back to the corresponding logic and fix whatever is missing.

On the spot it felt like a nice and quick improvement, but now I am much more concerned about the time I am stealing from you and myself on this task, so I'll close it! Anyone with a solid and cheap plan is free to reopen, of course. :)

Thanks again, @aaronrobertshaw!

**Comment by aaronrobertshaw** [comment]
> Looking at this with fresh eyes, I am now of the mind that this exploration is too costly for the payoff

I'm 100% on board with that assessment. Thanks all the same for taking the time to explore it though 🙇
